File: Algorithms/Sorting/Bucket_Sort/BucketSort.kt
fun insertion_sort(A: ArrayList<Double>) {
    var n = A.size
    var i: Int
    for (j in 1 until n) {
        var key = A[j]
        i = j - 1
        while (i >= 0 && A[i] > key) {
            A[i + 1] = A[i]
            i--
        }
        A[i + 1] = key
    }
}

fun bucket_sort(A:Array<Double>){
    var n=A.size
    var bucket = Array<ArrayList<Double>>(n,{i-> ArrayList() })   // creating buckets with n size
    for(i in 0..A.size-1){
        bucket[Math.floor(n*A[i]).toInt()].add(A[i])             // adding element a[i] into position n*a[i]
    }
    for(i in 0..A.size-1){
        insertion_sort(bucket[i])                               // sorting a list using inserton sort
    }
    for (i in 1..A.size-1){
        bucket[0].addAll(bucket[i])                              // concatenate all the buckets
    }
    println("After sorting :")
    for (i in bucket[0])
    {
        print("$i ")                                            // print the sorted elements
    }
}
fun main(arg: Array<String>) {
    print("Enter no. of elements :")
    var n = readLine()!!.toInt()

    println("Enter elements : ")
    var A = Array(n, { 0.0 })
    for (i in 0 until n)
        A[i] = readLine()!!.toDouble()

    bucket_sort(A)

}

File: Algorithms/Sorting/Bucket_Sort/BucketSort.js
const bucketSort = (arr, n = arr.length) => {
    //Create a bucket array
    let bucket = new Array(n);
    
    //Add bucket group
    for(let i = 0; i < n; i++){
      bucket[i] = [];
    }
    
    //Add the elements in a same range in bucket
    for(let i = 0; i < n; i++){
      let bucketIndex = Math.floor(arr[i]) * n;
      bucket[bucketIndex].push(arr[i]);
    }
    
    //Sort each bucket separately
    for(let i = 0; i < n; i++){
      bucket[i].sort();
    }
    
    // Get the sorted array
    let index = 0;
    for (let i = 0; i < n; i++) {
      for (let j = 0, size = bucket[i].length; j < size; j++) {
          arr[index++] = bucket[i][j];
      }
    }
  }